quinn-proto/src/connection/timer.rs

```diff
@@ -1,123 +1,356 @@
+        match timer {
+            Timer::Generic(timer) => {
+                self.generic[timer as usize] = Some(time);
+            }
+            Timer::PerPath(path_id, timer) => match self.path_timers.get_mut(&path_id) {
+                None => {
+                    let mut table = PathTimerTable::default();
+                    table.set(timer, time);
+                    self.path_timers.insert(path_id, table);
+                }
+                Some(table) => {
+                    table.set(timer, time);
+                }
+            },
+        }
     }
 
     pub(super) fn stop(&mut self, timer: Timer) {
-        self.most_recent_timeout.remove(&timer);
+        match timer {
+            Timer::Generic(timer) => {
+                self.generic[timer as usize] = None;
+            }
+            Timer::PerPath(path_id, timer) => {
+                if let Some(e) = self.path_timers.get_mut(&path_id) {
+                    e.stop(timer);
+                }
+            }
+        }
     }
 
     /// Get the next queued timeout
-    ///
-    /// Obsolete timers will be purged.
-    pub(super) fn peek(&mut self) -> Option<TimerEntry> {
-        while let Some(timer_entry) = self.timeout_queue.peek_mut() {
-            if self.most_recent_timeout.get(&timer_entry.timer) != Some(&timer_entry.time) {
-                // obsolete timeout
-                PeekMut::pop(timer_entry);
-                continue;
-            }
-            return Some(timer_entry.clone());
-        }
+    pub(super) fn peek(&mut self) -> Option<Instant> {
+        // TODO: this is currently linear in the number of paths
 
-        None
+        let min_generic = self.generic.iter().filter_map(|&x| x).min();
+        let min_path = self
+            .path_timers
+            .values()
+            .flat_map(|p| p.timers.iter().filter_map(|&x| x))
+            .min();
+
+        match (min_generic, min_path) {
+            (None, None) => None,
+            (Some(val), None) => Some(val),
+            (Some(a), Some(b)) => Some(a.min(b)),
+            (None, Some(val)) => Some(val),
+        }
     }
 
     /// Remove the next timer up until `now`, including it
-    pub(super) fn expire_before(&mut self, now: Instant) -> Option<Timer> {
-        let TimerEntry { time, timer } = self.peek()?;
-        if time <= now {
-            self.most_recent_timeout.remove(&timer);
-            self.timeout_queue.pop();
-            return Some(timer);
+    pub(super) fn expire_before(&mut self, now: Instant) -> Option<(Timer, Instant)> {
+        // TODO: this is currently linear in the number of paths
+
+        for timer in GenericTimer::VALUES {
+            if self.generic[timer as usize].is_some()
+                && self.generic[timer as usize].expect("checked") <= now
+            {
+                return self.generic[timer as usize]
+                    .take()
+                    .map(|time| (Timer::Generic(timer), time));
+            }
```

Comment:
I wonder if this could be a problem with preferring connection timers over path timers.  The old code and the code on main always prefers the timer which expired first, across all timers available.  Now under heavy load I think you can end up being erratic since the code no longer executes things in the right order.